src/Record.ts

```diff
@@ -624,32 +624,98 @@ export function filterWithKey<A>(fa: Record<string, A>, p: (key: string, a: A) =
 export function fromFoldable<F extends URIS2>(
   F: Foldable2<F>
-): <K extends string, L, A>(ta: Type2<F, L, [K, A]>, f: (existing: A, a: A) => A) => Record<K, A>
+): <K extends string, L, A>(ta: Type2<F, L, [K, A]>, concat: (existing: A, a: A) => A) => Record<K, A>
 export function fromFoldable<F extends URIS>(
   F: Foldable1<F>
-): <K extends string, A>(ta: Type<F, [K, A]>, f: (existing: A, a: A) => A) => Record<K, A>
+): <K extends string, A>(ta: Type<F, [K, A]>, concat: (existing: A, a: A) => A) => Record<K, A>
 export function fromFoldable<F>(
   // tslint:disable-next-line: deprecation
   F: Foldable<F>
-): <K extends string, A>(ta: HKT<F, [K, A]>, f: (existing: A, a: A) => A) => Record<K, A>
+): <K extends string, A>(ta: HKT<F, [K, A]>, concat: (existing: A, a: A) => A) => Record<K, A>
 export function fromFoldable<F>(
   // tslint:disable-next-line: deprecation
   F: Foldable<F>
-): <A>(ta: HKT<F, [string, A]>, f: (existing: A, a: A) => A) => Record<string, A> {
-  return <A>(ta: HKT<F, [string, A]>, f: (existing: A, a: A) => A) => {
-    return F.reduce<[string, A], Record<string, A>>(ta, {}, (b, [k, a]) => {
-      b[k] = b.hasOwnProperty(k) ? f(b[k], a) : a
-      return b
+): <A>(ta: HKT<F, [string, A]>, concat: (existing: A, a: A) => A) => Record<string, A> {
+  const fromFoldableMapF = fromFoldableMap(F)
+  return (ta, concat) => fromFoldableMapF(ta, identity, concat)
+}
+
+/**
+ * Create a record from a foldable collection using the specified functions to
+ *
+ * - map to key/value pairs
+ * - combine values for duplicate keys.
+ *
+ * @example
+ * import { array, zip } from 'fp-ts/lib/Array'
+ * import { identity } from 'fp-ts/lib/function'
+ * import { fromFoldableMap } from 'fp-ts/lib/Record'
+ *
+ * // like lodash `zipObject` or ramda `zipObj`
+ * export const zipObject = <K extends string, A>(keys: Array<K>, values: Array<A>): Record<K, A> =>
+ *   fromFoldableMap(array)(zip(keys, values), identity, (_, b) => b)
+ *
+ * assert.deepStrictEqual(zipObject(['a', 'b'], [1, 2, 3]), { a: 1, b: 2 })
+ *
+ * // build a map from a field
+ * interface User {
+ *   id: string
+ *   name: string
+ * }
+ *
+ * const users: Array<User> = [
+ *   { id: 'id1', name: 'name1' },
+ *   { id: 'id2', name: 'name2' },
+ *   { id: 'id1', name: 'name3' }
+ * ]
+ *
+ * assert.deepStrictEqual(fromFoldableMap(array)(users, user => [user.id, user], (_, b) => b), {
+ *   id1: { id: 'id1', name: 'name3' },
+ *   id2: { id: 'id2', name: 'name2' }
+ * })
+ *
+ * @since 1.16.0
+ */
+export function fromFoldableMap<F extends URIS3>(
+  F: Foldable3<F>
+): <U, L, A, K extends string, B>(
+  ta: Type3<F, U, L, A>,
+  f: (a: A) => [K, B],
+  concat: (existing: B, a: B) => B
+) => Record<K, B>
+export function fromFoldableMap<F extends URIS2>(
+  F: Foldable2<F>
+): <L, A, K extends string, B>(
+  ta: Type2<F, L, A>,
+  f: (a: A) => [K, B],
+  concat: (existing: B, a: B) => B
```

Comment:
I just curious, why we require here concat function instead of semigroup?

Reply:
@Wenqer mmmh admitely the name `concat` is misleading in this context, I'll change it to `onConflict`

> why we require here concat function instead of semigroup?

I'll answer with the opposite question: since requiring a `Semigroup` would add an additional constraint, why resolving conflicts should be associative? Is it possible to justify such a constraint?

Reply:
What about introducing `Magma`s? Afaics they don't require binary operation to be associative. This would make instance parametrization consistent across whole codebase and also make possible to parametrize `fromFoldableMap` with existing `Semigroup` instances.

Reply:
related https://github.com/gcanti/fp-ts/pull/431

Reply:
@raveclassic LGTM

Reply:
@gcanti I'll submit a PR then.